src/Bicep.Core/Emit/EmitLimitationCalculator.cs

```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
 
             return moduleScopeData.ToImmutableDictionary();
         }
+
+        private static IEnumerable<Diagnostic> DetectDuplicateNames(SemanticModel semanticModel, ImmutableDictionary<ResourceSymbol, ResourceSymbol?> resourceScopeData, ImmutableDictionary<ModuleSymbol, ScopeHelper.ScopeData> moduleScopeData)
+        {
+            // This method only checks, if in one deployment we do not have 2 or more resources with this same name in one deployment to avoid template validation error
+            // This will not check resource constraints such as necessity of having unique virtual network names within resource group
+
+            var duplicateResources = semanticModel.Root.AllDeclarations
+                .OfType<ResourceSymbol>()
+                .Where(res => res.Type is ResourceType && res.DeclaringResource.Body is ObjectSyntax)
+                .Select(res =>
+                {
+                    var namePropSyntax = ((ObjectSyntax)res.DeclaringResource.Body).SafeGetPropertyByName(LanguageConstants.ResourceNamePropertyName);
+                    var resourceScope = resourceScopeData[res];
+                    return new
+                    {
+                        ResourceName = res.Name,
+                        ResourceScope = resourceScope,
+                        ResourceTypeFQDN = ((ResourceType)res.Type).TypeReference.FullyQualifiedType,
+                        ResourceNamePropertyValue = namePropSyntax?.Value as StringSyntax, //TODO: currently limiting check to property values that are strings, although it can be references or other syntaxes
+                        ResourceNamePropertySyntax = namePropSyntax
+                    };
+                })
+                .Where(x => x.ResourceNamePropertySyntax is not null && x.ResourceNamePropertyValue is not null)
+                .GroupBy(x => (x.ResourceTypeFQDN, x.ResourceScope, x.ResourceNamePropertyValue!), ResourceComparer.Instance)
+                .Where(group => group.Count() > 1);
+
+            foreach (var duplicatedResourceGroup in duplicateResources)
+            {
+                var duplicatedResourceNames = string.Join(", ", duplicatedResourceGroup.Select(x => x.ResourceName));
```

Comment:
Please move list formatting logic into the DiagnosticBuilder method (see the use of `ToQuotedString` [here](https://github.com/Azure/bicep/blob/19140c30917d0129f84969f9070dd97110aded6b/src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs#L211) for an example)

Reply:
moved.